scripts/oneoff/export-approved-service-edits.py

```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env python3
+
+"""Get details about approved service edits in a period.
+
+Outputs a CSV.
+
+Usage: export-approved-service-edits.py [-h] [options] <from> [<to>]
+
+    <from>  Beginning of time range in ISO format.
+    <to>    End of time range in ISO format.
+
+Options:
+    --stage=<stage>  Stage to run script against [default: production].
+    -h, --help       Show this help text.
+
+Example:
+
+    Count approved service edits in October 2020
+
+    ./scripts/oneoff/export-approved-service-edits.py 2020-10 2020-11
+"""
+
+
+import csv
+import itertools
+import sys
+from datetime import datetime, timedelta
+from typing import Iterator, Optional
+
+from docopt import docopt
+
+from dmapiclient.data import DataAPIClient
+from dmapiclient.audit import AuditTypes
+from dmutils.email.helpers import validate_email_address
+from dmutils.env_helpers import get_api_endpoint_from_stage
+
+sys.path.insert(0, ".")
+
+from dmscripts.helpers.auth_helpers import get_auth_token
+
+
+ISO_FORMAT = "%Y-%m-%dT%H:%M:%S.%fZ"
+DEFAULT_DATETIME = "1970-01-01T00:00:00.000000Z"
+
+
+def parse_datetime(s: str) -> datetime:
+    try:
+        return datetime.strptime(s, ISO_FORMAT)
+    except ValueError as e:
+        if len(s) < len(DEFAULT_DATETIME):
+            return datetime.strptime(s + DEFAULT_DATETIME[len(s) :], ISO_FORMAT)
+        else:
+            raise e
+
+
+def find_approved_service_edits(
+    data_api_client, from_datetime: datetime, to_datetime: Optional[datetime] = None
+) -> Iterator[dict]:
+    """Find service edits approved by users in a period"""
+
+    audit_events = data_api_client.find_audit_events_iter(
+        acknowledged="true", latest_first=True, audit_type=AuditTypes.update_service
+    )
+
+    # use datetimes to make our life easier
+    audit_events = map(
+        lambda e: {
+            **e,
+            "acknowledgedAt": datetime.strptime(e["acknowledgedAt"], ISO_FORMAT),
+        },
+        audit_events,
+    )
+
+    if to_datetime:
+        audit_events = filter(lambda e: e["acknowledgedAt"] < to_datetime, audit_events)
+
+    # Audit events can be acknowledged at any point after they are created, and
+    # because the server response is sorted by `createdAt` this means in theory
+    # we would need to search every page of the response to find all
+    # acknowledged in a time period. In practice service edits tend to get approved
+    # by admins in fairly regular blocks, so we just fudge it a bit.
+    # TODO: add an option to the api to return results sorted by `acknowledgedAt`
```

Comment:
should we add a tech board card for this `TODO` so we don't forget about it?

Reply:
Good idea, done https://trello.com/c/EGinpnSg/1457-the-audit-events-api-endpoint-can-only-sort-audit-events-by-created-at